File: frontend/src/domains/auth/oidc-api.js
import {canBeReached} from "../../data/helper-functions.js";

const scopes = ['openid', 'email', 'profile', 'offline_access'];

export class OidcConfig {
  /**
   * @param clientId {string}
   * @param redirectUri {URL}
   * @param authorizationEndpoint {URL} oidc endpoint to authorize the client
   * @param tokenEndpoint {URL} oidc endpoint to get a token
   * @param userInfoEndpoint {URL} oidc endpoint to get info about the user
   * @param healthzEndpoint {URL} endpoint to check whether the oidc service is reachable
   * @param rolesKey {string} the key with which the roles of a user can be retrieved from the user info object.
   */
  constructor(
    clientId,
    redirectUri,
    authorizationEndpoint,
    tokenEndpoint,
    userInfoEndpoint,
    healthzEndpoint,
    rolesKey) {
    this.clientId = clientId;
    this.redirectUri = redirectUri;
    this.authorizationEndpoint = authorizationEndpoint;
    this.tokenEndpoint = tokenEndpoint;
    this.userInfoEndpoint = userInfoEndpoint;
    this.healthzEndpoint = healthzEndpoint;
    this.rolesKey = rolesKey;
  }
}

/**
 * Something the provider said no to.
 *
 * The status is kept because two of the answers mean different things to this
 * app: a provider that will not take a token is telling this device that the
 * token is no good, which is the device's to put right, and a provider that is
 * unwell is telling it nothing at all.
 */
export class OidcApiError extends Error {
  /**
   * @param message {string}
   * @param status {number}
   */
  constructor(message, status) {
    super(message);
    this.name = 'OidcApiError';
    this.status = status;
  }

  /**
   * Whether the provider refused what it was shown rather than failing to look
   * at it. A token that has run out, been revoked, or was signed with a key
   * that has since been rolled all land here.
   *
   * @return {boolean}
   */
  get isTokenRefused() {
    return this.status === 401 || this.status === 403;
  }
}

export class OidcApi {
  /**
   * @param oidcConfig {OidcConfig}
   */
  constructor(oidcConfig) {
    this._oidcConfig = oidcConfig;
  }

  async getFreshAccessToken() {
    let oidcFlowState = OidcStorage.oidcFlowState;
    if (oidcFlowState != null) {
      const urlParams = new URLSearchParams(window.location.search);
      const authorizationCode = urlParams.get('code');
      const receivedState = urlParams.get('state');
      if (authorizationCode !== null && authorizationCode.length !== 0 && oidcFlowState.state === receivedState) {
        // received callback from oidc server that an authorization which we started
        // has finished
        console.log('exchange authorization code for token');
        try {
          const tokenResponse = await OidcApi.callTokenEndpoint(
            this._oidcConfig.tokenEndpoint,
            TokenRequestParams.authorizationCode(this._oidcConfig.clientId, this._oidcConfig.redirectUri, authorizationCode, oidcFlowState.codeVerifier)
          );

          OidcStorage.tokenResponse = tokenResponse;
          OidcStorage.refreshToken = tokenResponse.refresh_token;
          return tokenResponse.access_token;
        } catch (e) {
          console.error('failed to exchange code for token', e);
          OidcStorage.tokenResponse = null;
          OidcStorage.refreshToken = null;
        } finally {
          OidcStorage.oidcFlowState = null;
        }
      }
    }

    const refreshToken = OidcStorage.refreshToken;
    if (refreshToken !== null) {
      console.log('refresh access token');
      try {
        const tokenResponse = await OidcApi.callTokenEndpoint(
          this._oidcConfig.tokenEndpoint,
          TokenRequestParams.refreshToken(this._oidcConfig.clientId, this._oidcConfig.redirectUri, scopes, refreshToken)
        );

        OidcStorage.tokenResponse = tokenResponse;
        OidcStorage.refreshToken = tokenResponse.refresh_token;
        return tokenResponse.access_token;
      } catch (e) {
        console.error('failed to exchange refresh token for token', e);
        OidcStorage.tokenResponse = null;
        OidcStorage.refreshToken = null;
      }
    }

    console.log('no callback happened or refresh-token received, initiating code flow');
    oidcFlowState = await OidcFlowState.Create();
    OidcStorage.oidcFlowState = oidcFlowState;
    OidcApi.navigateToAuthorizationEndpoint(this._oidcConfig, scopes, oidcFlowState);

    return null;
  }

  async getActiveAccessToken() {
    const tokenResponse = OidcStorage.tokenResponse;
    if (tokenResponse != null && tokenResponse.access_token != null) {
      return tokenResponse.access_token;
    }
    return await this.getFreshAccessToken();
  }

  /**
   * Asks the provider who is signed in.
   *
   * A token this app is holding can be refused however carefully it was dated:
   * a session ended at the provider, a key rolled, a clock that disagrees. So a
   * refusal is not an answer about the user, it is an answer about the token —
   * that one is thrown away and the question asked again with a fresh one,
   * which is what reloading the page used to do by accident.
   *
   * It is asked once more and no further. A provider that refuses a token it
   * has just issued is a provider that will go on refusing, and a loop between
   * the two of them is worse than being told.
   *
   * @return {Promise<UserInfoResponse>}
   */
  async getUserInfo() {
    const accessToken = await this.getActiveAccessToken();
    if (accessToken == null)
      throw new OidcApiError('failed to get access-token to getUserInfo', 401);

    try {
      return await OidcApi.callUserInfoEndpoint(
        this._oidcConfig.userInfoEndpoint,
        accessToken,
        this._oidcConfig.rolesKey
      );
    } catch (error) {
      if (!(error instanceof OidcApiError) || !error.isTokenRefused) {
        throw error;
      }

      console.log('the provider would not take the access token; asking for a new one');
      OidcStorage.tokenResponse = null;
      const freshAccessToken = await this.getFreshAccessToken();
      if (freshAccessToken == null) {
        // Either the provider has been navigated to and this page is on its way
        // out, or there was nothing to ask with. Neither is an answer.
        throw error;
      }

      return await OidcApi.callUserInfoEndpoint(
        this._oidcConfig.userInfoEndpoint,
        freshAccessToken,
        this._oidcConfig.rolesKey
      );
    }
  }

  /**
   * Whether the provider is there to be asked.
   *
   * A provider that cannot be reached is answered `false` rather than thrown
   * about: this is asked to find out whether to work from what is kept on the
   * device, and a network that is down is the very case it is asked in.
   *
   * @return {Promise<boolean>}
   */
  async canBeReached() {
    return await canBeReached(this._oidcConfig.healthzEndpoint);
  }

  /**
   * Makes a POST request to the token endpoint of the IDP and handles its
   * response.
   *
   * @param tokenEndpoint {URL} is the url to call.
   * @param authParams {TokenRequestParams} parameters to add to the request
   * @returns {Promise<TokenResponse>}
   */
  static async callTokenEndpoint(tokenEndpoint, authParams) {
    const response = await fetch(tokenEndpoint, {
      method: 'POST',
      headers: {'Content-Type': 'application/x-www-form-urlencoded'},
      body: authParams.toUrlSearchParams().toString(),
    });

    if (response.status >= 400) {
      throw new OidcApiError(
        `failed to get access-token: ${response.status} ${response.statusText}: ${await response.text()}`,
        response.status,
      );
    }

    return await response.json();
  }

  /**
   * @param userInfoEndpoint {URL}
   * @param accessToken {string}
   * @param rolesKey {string}
   * @return {Promise<UserInfoResponse>}
   */
  static async callUserInfoEndpoint(userInfoEndpoint, accessToken, rolesKey) {

    const response = await fetch(userInfoEndpoint, {
      method: 'GET',
      headers: {
        'Authorization': `Bearer ${accessToken}`
      }
    });

    if (response.status >= 400) {
      throw new OidcApiError(
        `failed to get user info: ${response.status} ${response.statusText}: ${await response.text()}`,
        response.status,
      );
    }

    const body = await response.json();
    return UserInfoResponse.fromResponse(body, rolesKey);
  }

  /**
   * @param oidcConfig {OidcConfig}
   * @param scopes {string[]}
   * @param oidcFlowState {OidcFlowState}
   */
  static navigateToAuthorizationEndpoint(oidcConfig, scopes, oidcFlowState) {
    const authParams = new URLSearchParams({
      client_id: oidcConfig.clientId,
      redirect_uri: oidcConfig.redirectUri.toString(),
      scope: scopes.join(' '),
      response_type: 'code',
      code_challenge: oidcFlowState.codeChallenge,
      code_challenge_method: oidcFlowState.codeChallengeMethod,
      state: oidcFlowState.state
    })

    const authUrl = `${oidcConfig.authorizationEndpoint.protocol}//${oidcConfig.authorizationEndpoint.host}${oidcConfig.authorizationEndpoint.pathname}?${authParams.toString()}`;
    console.log(`navigating for auth: ${authUrl}`);
    window.location.href = authUrl;

    // return false needs to be here to make the redirect work...
    return false;
  }
}

/**
 * @param {number} length is the length of requested string
 * @returns {string} A random string ([A-Za-z0-9]+)
 */
function generateRandomString(length) {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  const array = new Uint8Array(length);
  crypto.getRandomValues(array);
  return Array
    .from(array, n => chars[n % chars.length])
    .join('');
}

/**
 * Hashes a given input string as a SHA256. This function exists because the
 * crypto functionality of a browser does not exist inside an "insecure context"
 * like our dev-evn.
 *
 * @param input {string}
 * @return {Promise<string>} the digest as a binary string, one character per
 *   byte, which is what btoa expects.
 */
async function sha256(input) {
  if (isSecureContext) {
    const buffer = new TextEncoder().encode(input);
    const hashBuff = await crypto.subtle.digest('SHA-256', buffer);
    return String.fromCharCode(...new Uint8Array(hashBuff));
  }

  console.warn('Running in insecure context. Using CryptoJS instead of browser built-in');
  if (!window.CryptoJS) {
    await new Promise((resolve, reject) => {
      const s = document.createElement('script');
      s.src = 'https://cdn.jsdelivr.net/npm/crypto-js@4.2.0/crypto-js.min.js';
      s.onload = resolve;
      s.onerror = reject;
      document.head.appendChild(s);
    });
  }
  return window.CryptoJS.SHA256(input).toString(window.CryptoJS.enc.Latin1);
}


// ----------------------------------------------------------------------------
// MODELS
// ----------------------------------------------------------------------------

export class OidcFlowState {
  /**
   *
   * @param state {string}
   * @param codeVerifier {string}
   * @param codeChallenge {string}
   * @param codeChallengeMethod {string}
   */
  constructor(state, codeVerifier, codeChallenge, codeChallengeMethod) {
    this.state = state;
    this.codeVerifier = codeVerifier;
    this.codeChallenge = codeChallenge;
    this.codeChallengeMethod = codeChallengeMethod;
  }

  /**
   * @return {Promise<OidcFlowState>}
   */
  static async Create() {
    const state = generateRandomString(16);
    const codeVerifier = generateRandomString(56);
    const codeChallenge = await OidcFlowState.createCodeChallenge(codeVerifier);

    return new OidcFlowState(state, codeVerifier, codeChallenge, 'S256');
  }

  /**
   * @param {string} verifier is the verifier which should be used for the challenge
   * @returns {Promise<string>} The SHA256 hashed input.
   */
  static async createCodeChallenge(verifier) {
    const verifierHash = await sha256(verifier)

    return btoa(verifierHash) // base64 encode
      .replace(/\+/g, '-')
      .replace(/\//g, '_')
      .replace(/=+$/, '');
  }
}

/**
 * Response from the token endpoint after requesting a token from the IDP.
 */
export class TokenResponse {
  /**
   * @param access_token {string}
   * @param refresh_token {string|null}
   * @param expires_in {number}
   * @param id_token {string}
   * @param token_type {string}
   */
  constructor(
    access_token,
    refresh_token,
    expires_in,
    id_token,
    token_type,
  ) {
    this.access_token = access_token;
    this.refresh_token = refresh_token;
    this.expires_in = expires_in;
    this.id_token = id_token;
    this.token_type = token_type;
  }
}

/**
 * Request with which a token can be requested from the IDP.
 */
export class TokenRequestParams {
  /**
   * @param client_id {string}
   * @param grant_type {"refresh_token"|"authorization_code"}
   * @param redirect_uri {URL}
   * @param scope {string[]|null}
   * @param refresh_token {string|null}
   * @param code {string|null}
   * @param code_verifier {string|null}
   */
  constructor(
    client_id,
    grant_type,
    redirect_uri,
    scope,
    refresh_token,
    code,
    code_verifier) {
    this.clientId = client_id;
    this.grantType = grant_type;
    this.redirectUri = redirect_uri;
    this.scope = scope;
    this.refreshToken = refresh_token;
    this.code = code;
    this.codeVerifier = code_verifier;
  }

  /**
   * @param clientId {string}
   * @param redirectUrl {URL}
   * @param code {string}
   * @param codeVerifier {string}
   */
  static authorizationCode(clientId, redirectUrl, code, codeVerifier) {
    return new TokenRequestParams(clientId, 'authorization_code', redirectUrl, null, null, code, codeVerifier);
  }

  /**
   * @param clientId {string}
   * @param redirectUrl {URL}
   * @param scope {string[]}
   * @param refreshToken {string}
   */
  static refreshToken(clientId, redirectUrl, scope, refreshToken) {
    return new TokenRequestParams(clientId, 'refresh_token', redirectUrl, scope, refreshToken, null, null);
  }

  toUrlSearchParams() {
    return new URLSearchParams({
      client_id: this.clientId,
      grant_type: this.grantType,
      redirect_uri: this.redirectUri.toString(),
      scope: this.scope?.join(' '),
      refresh_token: this.refreshToken,
      code: this.code,
      code_verifier: this.codeVerifier,
    });
  }
}

export class UserInfoResponse {
  /**
   * Which claims a user-info response carries is up to the provider, so every
   * one of these can be absent.
   *
   * The claims it was read out of are kept alongside it. What a provider
   * answers with is the one thing that explains why this app thinks what it
   * thinks about a user — which is worth being able to show when it thinks
   * something the user disagrees with.
   *
   * @param name {string|null}
   * @param subject {string|null}
   * @param email {string|null}
   * @param roles {Object|null}
   * @param claims {Object|null} the answer this was read out of
   * @param rolesKey {string|null} the claim the roles were looked for under
   */
  constructor(name, subject, email, roles, claims = null, rolesKey = null) {
    this.name = name;
    this.subject = subject;
    this.email = email;
    this.roles = roles;
    this.claims = claims;
    this.rolesKey = rolesKey;
  }

  /**
   * @return {boolean}
   */
  get isScoreEditor() {
    return this.roles != null && this.roles['score_editor'] != null;
  }

  get isScoreViewer() {
    return this.roles != null && this.roles['score_viewer'] != null;
  }

  /**
   * @param response {Object}
   * @param rolesKey {string}
   * @return {UserInfoResponse}
   */
  static fromResponse(response, rolesKey) {
    // A claim the provider did not send is absent, not the text "undefined",
    // which is what concatenating an empty string onto it would produce.
    return new UserInfoResponse(
      response['name'] ?? null,
      response['sub'] ?? null,
      response['email'] ?? null,
      response[rolesKey] ?? null,
      response,
      rolesKey
    );
  }

  /**
   * The same user, read back from what {@link JSON.stringify} made of one.
   *
   * What comes out of `JSON.parse` is a plain object: it carries the fields but
   * none of the methods, so `isScoreViewer` on it is not `false` but `undefined`
   * — and every question this app asks about a user is asked that way. Reading
   * it back into a user is what keeps a user who is kept on the device from
   * being a user with no roles at all.
   *
   * @param json {Object|null}
   * @return {UserInfoResponse|null}
   */
  static fromJson(json) {
    if (json == null || typeof json !== 'object') {
      return null;
    }
    return new UserInfoResponse(
      json.name ?? null,
      json.subject ?? null,
      json.email ?? null,
      json.roles ?? null,
      json.claims ?? null,
      json.rolesKey ?? null
    );
  }
}

const oidcFlowStateSessionStorageKey = 'auth_oidc_flow_state';
const tokenResponseSessionStorageKey = 'auth_token_response';
const refreshTokenSessionStorageKey = 'auth_refresh_token';

/**
 * How long before it actually runs out a token is treated as run out. A request
 * carrying a token with two seconds left is a request that arrives without one.
 *
 * @type {number}
 */
const tokenExpiryMargin = 30 * 1000;

/**
 * Writes a value to the session storage, or clears the key when there is no
 * value. Storing null is not the same as clearing: setItem stringifies, so it
 * would leave the string "null" behind for the next reader to trip over.
 *
 * @param key {string}
 * @param value {string|null|undefined}
 */
function setOrRemove(key, value) {
  if (value == null) {
    sessionStorage.removeItem(key);
    return;
  }
  sessionStorage.setItem(key, value);
}

export class OidcStorage {

  /**
   * @param value {TokenResponse|null}
   */
  static set tokenResponse(value) {
    if (value == null) {
      setOrRemove(tokenResponseSessionStorageKey, null);
      return;
    }
    // When it runs out, rather than how long it had left when it arrived.
    //
    // This used to be a timer that cleared the token when it expired, and a
    // timer cannot say this: it dies with the page. A tab closed with five
    // minutes left on a token and opened again an hour later found the token
    // still sitting in the storage with nothing left to clear it, handed it to
    // the provider, and was told 401 — which is a thing this app can now say
    // for itself, before asking anybody.
    setOrRemove(tokenResponseSessionStorageKey, JSON.stringify({
      ...value,
      // A provider that did not say when is a token this app cannot date, and
      // one it will not throw away on a guess. It finds out the way it always
      // did: by being refused.
      expires_at: value.expires_in > 0 ? Date.now() + value.expires_in * 1000 : null,
    }));
  }

  /**
   * @return {TokenResponse|null}
   */
  static get tokenResponse() {
    const json = sessionStorage.getItem(tokenResponseSessionStorageKey);
    if (json == null) {
      return null;
    }

    const tokenResponse = JSON.parse(json);
    if (tokenResponse?.expires_at == null) {
      return tokenResponse;
    }

    // Given up a little before it is actually out, so that a token cannot run
    // out somewhere between being read here and arriving at the server.
    if (Date.now() < tokenResponse.expires_at - tokenExpiryMargin) {
      return tokenResponse;
    }
    OidcStorage.tokenResponse = null;
    return null;
  }

  /**
   * @param value {OidcFlowState|null}
   */
  static set oidcFlowState(value) {
    setOrRemove(oidcFlowStateSessionStorageKey, value == null ? null : JSON.stringify(value));
  }

  /**
   * @return {OidcFlowState|null}
   */
  static get oidcFlowState() {
    const json = sessionStorage.getItem(oidcFlowStateSessionStorageKey);
    if (json == null) {
      return null;
    }
    return JSON.parse(json);
  }

  /**
   * @param value {string|null}
   */
  static set refreshToken(value) {
    setOrRemove(refreshTokenSessionStorageKey, value);
  }

  /**
   * @return {string|null}
   */
  static get refreshToken() {
    return sessionStorage.getItem(refreshTokenSessionStorageKey);
  }
}

File: frontend_flutter/lib/domains/auth/oidc_api.dart
import 'dart:convert';
import 'dart:math';

import 'package:crypto/crypto.dart';
import 'package:http/http.dart' as http;

import '../../config.dart';
import '../../data/local_store.dart';
import 'authorizer.dart';

/// Proving who the user is, and finding out what they may do.
///
/// The flow is an authorization code with PKCE, which is the one a client that
/// cannot keep a secret is allowed to use — and a browser, a phone and a laptop
/// are all clients that cannot keep a secret. What the app gets out of it is an
/// access token to call the API with and a set of roles to decide what to show.
///
/// Where this differs from the app it replaces: the tokens are kept where they
/// survive a restart rather than only for as long as a tab is open. A device
/// that is closed and opened again at the next rehearsal should not ask the
/// player to sign in again, and a refresh token that lives no longer than a tab
/// is a refresh token that never gets used. Signing out throws them away, which
/// is what the profile page is for.
const List<String> _scopes = ['openid', 'email', 'profile', 'offline_access'];

const _tokenKey = 'auth_token_response';
const _refreshTokenKey = 'auth_refresh_token';
const _flowStateKey = 'auth_flow_state';
const _userInfoKey = 'app_user_info';

class OidcApi {
  OidcApi(this._config, this._store, {http.Client? client, Authorizer? authorizer})
      : _client = client ?? http.Client(),
        _authorizer = authorizer ?? Authorizer(_config);

  final OidcConfig _config;
  final LocalStore _store;
  final http.Client _client;
  final Authorizer _authorizer;

  /// A token that is good right now, asking for a new one when there is not
  /// one. `null` when the user has to sign in and the app is about to send them
  /// away to do it.
  Future<String?> getActiveAccessToken() async {
    final held = await _heldToken();
    if (held != null) {
      return held;
    }
    return getFreshAccessToken();
  }

  Future<String?> _heldToken() async {
    final json = await _store.readSetting(_tokenKey);
    if (json == null) {
      return null;
    }

    final token = jsonDecode(json) as Map<String, dynamic>;
    final expiresAt = token['expires_at'];
    if (expiresAt is int &&
        DateTime.now().millisecondsSinceEpoch >
            expiresAt - const Duration(seconds: 30).inMilliseconds) {
      // Half a minute of slack, so that a token is never spent on a request
      // that will take longer to arrive than the token has left.
      return null;
    }
    final access = token['access_token'];
    return access is String && access.isNotEmpty ? access : null;
  }

  /// Gets a token by whatever means are left: the code the user just came back
  /// with, the refresh token this device is holding, or by asking them.
  Future<String?> getFreshAccessToken() async {
    final callback = await _authorizer.pendingCallback();
    if (callback != null) {
      final token = await _exchangeCallback(callback);
      await _authorizer.clearCallback();
      if (token != null) {
        return token;
      }
    }

    final refreshToken = await _store.readSetting(_refreshTokenKey);
    if (refreshToken != null) {
      final token = await _refresh(refreshToken);
      if (token != null) {
        return token;
      }
    }

    return _startFlow();
  }

  /// Turns the code the user came back with into a token. `null` when there was
  /// nothing to turn, or when the provider refused it.
  Future<String?> _exchangeCallback(Callback callback) async {
    final flow = await _readFlowState();
    if (flow == null || flow.state != callback.state) {
      // A code that came back with a state this device never sent is a code
      // this device did not ask for.
      return null;
    }

    try {
      final token = await _callTokenEndpoint({
        'client_id': _config.clientId,
        'grant_type': 'authorization_code',
        'redirect_uri': _authorizer.redirectUri.toString(),
        'code': callback.code,
        'code_verifier': flow.verifier,
      });
      await _keep(token);
      return token['access_token'] as String?;
    } catch (error) {
      await _forgetTokens();
      return null;
    } finally {
      await _store.writeSetting(_flowStateKey, null);
    }
  }

  Future<String?> _refresh(String refreshToken) async {
    try {
      final token = await _callTokenEndpoint({
        'client_id': _config.clientId,
        'grant_type': 'refresh_token',
        'redirect_uri': _authorizer.redirectUri.toString(),
        'scope': _scopes.join(' '),
        'refresh_token': refreshToken,
      });
      await _keep(token);
      return token['access_token'] as String?;
    } catch (error) {
      await _forgetTokens();
      return null;
    }
  }

  /// Sends the user to the provider. On a device the answer comes back here; on
  /// the web the app is on its way out and the answer will be waiting when it
  /// starts again.
  Future<String?> _startFlow() async {
    final flow = _FlowState.create();
    await _store.writeSetting(_flowStateKey, jsonEncode(flow.toJson()));

    final url = _config.authorizationEndpoint.replace(queryParameters: {
      'client_id': _config.clientId,
      'redirect_uri': _authorizer.redirectUri.toString(),
      'scope': _scopes.join(' '),
      'response_type': 'code',
      'code_challenge': flow.challenge,
      'code_challenge_method': 'S256',
      'state': flow.state,
    });

    final callback = await _authorizer.authorize(url);
    if (callback == null) {
      return null;
    }
    return _exchangeCallback(callback);
  }

  Future<Map<String, dynamic>> _callTokenEndpoint(
      Map<String, String> body) async {
    final response = await _client.post(
      _config.tokenEndpoint,
      headers: const {'Content-Type': 'application/x-www-form-urlencoded'},
      body: body,
    );

    if (response.statusCode >= 400) {
      throw OidcException(
        'failed to get an access token: ${response.statusCode} ${response.body}',
      );
    }
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  /// Keeps what the provider sent, with the moment it runs out worked out now:
  /// what it says is how long the token lasts, and how long is only an answer
  /// while it is being read.
  Future<void> _keep(Map<String, dynamic> token) async {
    final expiresIn = token['expires_in'];
    token['expires_at'] = DateTime.now()
        .add(Duration(seconds: expiresIn is int ? expiresIn : 300))
        .millisecondsSinceEpoch;

    await _store.writeSetting(_tokenKey, jsonEncode(token));

    final refresh = token['refresh_token'];
    if (refresh is String && refresh.isNotEmpty) {
      await _store.writeSetting(_refreshTokenKey, refresh);
    }
  }

  Future<void> _forgetTokens() async {
    await _store.writeSetting(_tokenKey, null);
    await _store.writeSetting(_refreshTokenKey, null);
  }

  Future<_FlowState?> _readFlowState() async {
    final json = await _store.readSetting(_flowStateKey);
    if (json == null) {
      return null;
    }
    return _FlowState.fromJson(jsonDecode(json) as Map<String, dynamic>);
  }

  /// What the provider says about the user right now.
  Future<UserInfo?> getUserInfo() async {
    final token = await getActiveAccessToken();
    if (token == null) {
      return null;
    }

    final response = await _client.get(
      _config.userInfoEndpoint,
      headers: {'Authorization': 'Bearer $token'},
    );
    if (response.statusCode >= 400) {
      throw OidcException(
        'failed to get the user info: ${response.statusCode} ${response.body}',
      );
    }

    final claims = jsonDecode(response.body) as Map<String, dynamic>;
    final user = UserInfo.fromClaims(claims, _config.rolesKey);
    await _store.writeSetting(_userInfoKey, jsonEncode(user.toJson()));
    return user;
  }

  /// What the provider last said about the user, from before there was no
  /// network to ask over.
  Future<UserInfo?> keptUserInfo() async {
    final json = await _store.readSetting(_userInfoKey);
    if (json == null) {
      return null;
    }
    return UserInfo.fromJson(jsonDecode(json) as Map<String, dynamic>);
  }

  /// Whether the provider is there to be asked.
  ///
  /// A provider that cannot be reached is answered `false` rather than thrown
  /// about: this is asked to find out whether to work from what is kept on the
  /// device, and a network that is down is the very case it is asked in.
  Future<bool> canBeReached() async {
    try {
      final response = await _client
          .get(_config.healthzEndpoint)
          .timeout(const Duration(seconds: 5));
      return response.statusCode < 400;
    } catch (error) {
      return false;
    }
  }

  /// Forgets who is signed in on this device.
  ///
  /// It signs nobody out at the provider — that is the provider's own business,
  /// and this app is in no position to speak for it. What it does is make the
  /// next visit ask again from the beginning, which is the way out of a token
  /// or a set of roles that has gone stale.
  ///
  /// The scores and sets on this device are left alone: they are what makes the
  /// app work without a network, and they are no use to anyone who cannot get a
  /// token to read them with anyway.
  Future<void> forgetUser() async {
    await _forgetTokens();
    await _store.writeSetting(_flowStateKey, null);
    await _store.writeSetting(_userInfoKey, null);
  }
}

class OidcException implements Exception {
  OidcException(this.message);

  final String message;

  @override
  String toString() => message;
}

// ---------------------------------------------------------------------------
// THE FLOW
// ---------------------------------------------------------------------------

/// What this device has to remember while the user is away at the provider: the
/// secret it will prove the code with, and a nonce to recognise its own answer
/// by.
class _FlowState {
  _FlowState(this.state, this.verifier, this.challenge);

  final String state;
  final String verifier;
  final String challenge;

  static _FlowState create() {
    final verifier = _randomString(56);
    return _FlowState(_randomString(16), verifier, _challengeFor(verifier));
  }

  Map<String, dynamic> toJson() =>
      {'state': state, 'verifier': verifier, 'challenge': challenge};

  factory _FlowState.fromJson(Map<String, dynamic> json) => _FlowState(
        '${json['state']}',
        '${json['verifier']}',
        '${json['challenge']}',
      );
}

const _alphabet =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

String _randomString(int length) {
  final random = Random.secure();
  return String.fromCharCodes([
    for (var i = 0; i < length; i++)
      _alphabet.codeUnitAt(random.nextInt(_alphabet.length)),
  ]);
}

/// The verifier as the provider will check it: hashed, and written the way a
/// url can carry it.
String _challengeFor(String verifier) => base64Url
    .encode(sha256.convert(utf8.encode(verifier)).bytes)
    .replaceAll('=', '');

// ---------------------------------------------------------------------------
// THE USER
// ---------------------------------------------------------------------------

/// What the provider said about the user, and what this app made of it.
///
/// The claims it was read out of are kept alongside it. What a provider answers
/// with is the one thing that explains why this app thinks what it thinks about
/// a user — which is worth being able to show when it thinks something the user
/// disagrees with.
class UserInfo {
  const UserInfo({
    this.name,
    this.subject,
    this.email,
    this.roles,
    this.claims,
    this.rolesKey,
  });

  final String? name;
  final String? subject;
  final String? email;

  /// The roles as the provider sent them, which is a map whose keys are the
  /// role names.
  final Map<String, dynamic>? roles;

  /// The answer this was read out of.
  final Map<String, dynamic>? claims;

  /// The claim the roles were looked for under.
  final String? rolesKey;

  bool get isScoreEditor => roles?['score_editor'] != null;

  bool get isScoreViewer => roles?['score_viewer'] != null;

  factory UserInfo.fromClaims(Map<String, dynamic> claims, String rolesKey) {
    final roles = claims[rolesKey];
    return UserInfo(
      name: claims['name'] as String?,
      subject: claims['sub'] as String?,
      email: claims['email'] as String?,
      roles: roles is Map<String, dynamic> ? roles : null,
      claims: claims,
      rolesKey: rolesKey,
    );
  }

  Map<String, dynamic> toJson() => {
        'name': name,
        'subject': subject,
        'email': email,
        'roles': roles,
        'claims': claims,
        'rolesKey': rolesKey,
      };

  factory UserInfo.fromJson(Map<String, dynamic> json) => UserInfo(
        name: json['name'] as String?,
        subject: json['subject'] as String?,
        email: json['email'] as String?,
        roles: (json['roles'] as Map?)?.cast<String, dynamic>(),
        claims: (json['claims'] as Map?)?.cast<String, dynamic>(),
        rolesKey: json['rolesKey'] as String?,
      );
}
